### PR TITLE
Servo gimbal

### PR DESCRIPTION
**Pull request by mje-nz**
Implements #8120.

In the end I did a range and offset parameter for each channel, which is enough to fully tune a servo gimbal without touching the mixer settings.

I've tested this with a single-axis gimbal. It works as expected with EKF2, and for some reason I have to manually launch attitude_estimator_q at the moment but once I do it also works fine. There should be no change in behavior with the default parameters (apart from the de-optimization around parameter updates, which should not cause a visible change in behavior).

### CODE REVIEW

File: src/drivers/vmount/input_rc.h [hunk 1]
```diff
@@ -80,6 +81,7 @@ class InputRC : public InputBase
 	float _get_aux_value(const manual_control_setpoint_s &manual_control_setpoint, int channel_idx);
 
 private:
+	bool _do_stabilization;
```
Comment (bkueng) [inline src/drivers/vmount/input_rc.h]: This can be `const`

File: src/drivers/vmount/output.h [hunk 2]
```diff
@@ -53,6 +53,13 @@ struct OutputConfig {
 	float gimbal_retracted_mode_value; /**< mixer output value for selecting gimbal retracted mode */
 	float gimbal_normal_mode_value; /**< mixer output value for selecting gimbal normal mode */
 
+	float pitch_range;
```
Comment (bkueng) [inline src/drivers/vmount/output.h]: Please store these in radians

File: src/drivers/vmount/vmount_params.c [hunk 3]
```diff
@@ -156,3 +157,73 @@ PARAM_DEFINE_INT32(MNT_MAN_PITCH, 0);
 */
 PARAM_DEFINE_INT32(MNT_MAN_YAW, 0);
 
+/**
+* Stabilize the mount (enable for servo gimbal, disable for passthrough).
+*
+* @value 0 Disable
+* @value 1 Enable
+* @min 0
+* @max 1
```
Comment (bkueng) [inline src/drivers/vmount/vmount_params.c]: You can use `@boolean` instead.

File: src/drivers/vmount/output_rc.cpp [hunk 4]
```diff
@@ -73,9 +73,13 @@ int OutputRC::update(const ControlData *control_data)
 
 	actuator_controls_s actuator_controls;
 	actuator_controls.timestamp = hrt_absolute_time();
-	actuator_controls.control[0] = _angle_outputs[0] / M_PI_F;
-	actuator_controls.control[1] = _angle_outputs[1] / M_PI_F;
-	actuator_controls.control[2] = _angle_outputs[2] / M_PI_F;
+	// _angle_outputs are in radians, actuator_controls are in [-1, 1]
+	actuator_controls.control[0] = (_angle_outputs[0] + _config.roll_offset * M_DEG_TO_RAD_F) /
+				       (_config.roll_range * (M_DEG_TO_RAD_F / 2.0f));
```
Comment (ndepal) [inline src/drivers/vmount/output_rc.cpp]: why `/2.0f`?

File: src/drivers/vmount/input_rc.h (hunk 1 above)
Comment (mje-nz) [reply src/drivers/vmount/input_rc.h]: Thanks

File: src/drivers/vmount/output.h (hunk 2 above)
Comment (mje-nz) [reply src/drivers/vmount/output.h]: Is that an optimization or a convention? If it's an optimization, how about storing pitch_scale = 1/(pitch_range * (M_DEG_TO_RAD_F / 2.0f)) instead?

File: src/drivers/vmount/vmount_params.c (hunk 3 above)
Comment (mje-nz) [reply src/drivers/vmount/vmount_params.c]: Thanks

File: src/drivers/vmount/output_rc.cpp (hunk 4 above)
Comment (mje-nz) [reply src/drivers/vmount/output_rc.cpp]: The purpose of the angle range parameters is to generalise this to support servos, rather than just outputting into brushless gimbal controllers. A standard servo has a movement range of somewhere between 60 and 180 degrees from MIN_PWM to MAX_PWM, so this control output calculation needs to take that into account. See the Ardupilot documentation of their version [here](http://ardupilot.org/copter/docs/common-camera-gimbal.html).

For the factor of 2 specifically: for a 90 degree servo this expression needs to be a linear transformation from [-45, 45] to [-1, 1] (ignoring the offset), hence dividing by range/2.

File: src/drivers/vmount/vmount.cpp [hunk 5]
```diff
@@ -228,10 +230,16 @@ static int vmount_thread_main(int argc, char *argv[])
 
 		if (!thread_data.input_objs[0] && (params.mnt_mode_in >= 0 || test_input)) { //need to initialize
 
-			output_config.gimbal_normal_mode_value = params.mnt_ob_norm_mode;
-			output_config.gimbal_retracted_mode_value = params.mnt_ob_lock_mode;
-			output_config.mavlink_sys_id = params.mnt_mav_sysid;
-			output_config.mavlink_comp_id = params.mnt_mav_compid;
+            output_config.gimbal_normal_mode_value = params.mnt_ob_norm_mode;
```
[the remark this answers is not in the document]
Comment (mje-nz) [reply src/drivers/vmount/vmount.cpp]: Damn it, missed one. Thanks.

File: src/drivers/vmount/vmount_params.c [hunk 6]
```diff
@@ -156,3 +157,73 @@ PARAM_DEFINE_INT32(MNT_MAN_PITCH, 0);
 */
 PARAM_DEFINE_INT32(MNT_MAN_YAW, 0);
 
+/**
+* Stabilize the mount (enable for servo gimbal, disable for passthrough).
+*
```
[the remark this answers is not in the document]
Comment (mje-nz) [reply src/drivers/vmount/vmount_params.c]: Thanks.

File: src/drivers/vmount/output_rc.cpp (hunk 4 above)
Comment (ndepal) [reply src/drivers/vmount/output_rc.cpp]: I see, thanks! How problematic is it if you output something that's outside of that range? Is it worth guarding against ever having `actuator_controls.control[i]` > 1 or <-1?

File: src/drivers/vmount/output.h (hunk 2 above)
Comment (bkueng) [reply src/drivers/vmount/output.h]: Convention: we typically use rad for rotation angles. If you want to further optimize, that's fine with me, just make sure the meaning is clearly documented.

File: src/drivers/vmount/output_rc.cpp (hunk 4 above)
Comment (mje-nz) [reply src/drivers/vmount/output_rc.cpp]: Won't the mixer saturate at 1 and -1?

What happens to a servo at and beyond the edge of its range seems to vary; of the two I tested, an expensive one saturated on its own and a cheap one jerked around.

It's certainly possible that someone could want to set upper and lower limits on a particular channel (e.g. to prevent a camera from pointing back at the frame, or if there are clearance constraints). I feel it's probably fine to just expect people to tweak their mixers for that though, since it's an airframe-dependent setting you only have to set once. The parameters I've added are the ones you need to iteratively tweak until the gimbal stabilizes correctly, so it would be quite inconvenient to do that with mixers. What do you think?

File: src/drivers/vmount/output.h (hunk 2 above)
Comment (mje-nz) [reply src/drivers/vmount/output.h]: Fair enough. Are you happy with the parameters and cached parameter values being in degrees?

File: src/drivers/vmount/vmount_params.c [hunk 7]
```diff
@@ -156,3 +157,71 @@ PARAM_DEFINE_INT32(MNT_MAN_PITCH, 0);
 */
 PARAM_DEFINE_INT32(MNT_MAN_YAW, 0);
 
+/**
+* Stabilize the mount (true for servo gimbal, false for passthrough).
+* Only affects RC input; Mavlink input can set this with the DO_MOUNT_CONFIGURE command.
```
Comment (mje-nz) [inline src/drivers/vmount/vmount_params.c]: Actually, how about this parameter also sets the default on InputMavlinkCmdMount for consistency? (@ndepal)

File: src/drivers/vmount/output.h (hunk 2 above)
Comment (bkueng) [reply src/drivers/vmount/output.h]: Yes, since the parameters are user-facing.

File: src/drivers/vmount/vmount_params.c (hunk 7 above)
Comment (ndepal) [reply src/drivers/vmount/vmount_params.c]: Good idea.

**Comment by mje-nz** [comment]
@ndepal: I've tested it with both RC and MavlinkCmdMount input to make sure that last commit worked.
@bkueng: Is that everything, aside from the parameter update optimization you asked @dagar about?

File: src/drivers/vmount/output_mavlink.cpp [hunk 8]
```diff
@@ -113,9 +113,9 @@ int OutputMavlink::update(const ControlData *control_data)
 
 	// vmount spec has roll, pitch on channels 0, 1, respectively; MAVLink spec has roll, pitch on channels 1, 0, respectively
 	// vmount uses radians, MAVLink uses degrees
-	vehicle_command.param1 = _angle_outputs[1] * M_RAD_TO_DEG_F;
```
Comment (bkueng) [inline src/drivers/vmount/output_mavlink.cpp]: Why did you remove the `* M_RAD_TO_DEG_F` factor?

Comment (mje-nz) [reply src/drivers/vmount/output_mavlink.cpp]: Oops, looks like I got a bit overenthusiastic when I changed the offset to be in radians, thanks.

Comment (mje-nz) [reply src/drivers/vmount/output_mavlink.cpp]: (resolved)

**Comment by mje-nz** [comment]
@bkueng Ok, I've added your commit, and put the parameter check back in (as you requested in your first comment). I've put your commit at the tip so they're in the right order chronologically, would you prefer it before mine so the intermediate commits build? If not, this should be ready to merge now.

**Comment by bkueng** [comment]
Jenkins CI failure is unrelated

**Comment by mje-nz** [comment]
Thanks for your help :)

**Comment by mzahana** [comment]
Hi. So, can a servo gimbal be automatically stabilized after this merge? If yes, how does it work?
Thanks

**Comment by mje-nz** [comment]
@mzahana Yes: Hook up the gimbal as normal, turn on MNT_DO_STAB, use MNT_RANGE_x to get each servo's range correct, and use MNT_OFF_x to get each servo's centre correct.

**Comment by mzahana** [comment]
@mje-nz Thanks. It would be very nice if there is a separate window in QGC to set those parameters like in mission planner.

**Comment by mje-nz** [comment]
@mzahana I agree, and I was thinking of that when I implemented the range and offset parameters. It's very far down my to-do list though, so feel free to have a go yourself!

**Comment by ryanjAA** [comment]
I posted this on slack but thought this might be a better place since you wrote it. I can’t seem to get the vmount working. I have a servo gimbal plugged in to aux 1 and 2 (just servos plugged in for testing), mnt_mode_in as RC mnt_mode_out as AUX and mnt_so_stab as enabled. I don’t see any movement on the servos though. I have the flight controller armed as well and powered via the servo rail to rule out any of that. Am i missing something? I’m just trying to use the vmount to stabilize two servos and use the Pixhawk as a means for stabilizing them.

**Comment by mje-nz** [comment]
A two-year-old closed pull request isn't really an appropriate place to ask for help, but here's a few thing to try anyway:

* Do you have the safety switch off?
* Do you have `MNT_MAN_PITCH`/`MNT_MAN_ROLL` etc set?
* Can you get the servos to move using `vmount test` in the console?
* Depending on your flight controller, try main 5/6 instead of aux 1/2.

If none of that helps, please open an issue :)
